File: ios/Sources/NoteMeldMobile/MobileSurfaceState.swift
import Foundation

/// Shared page contract for the iOS host. Agent execution stays in the SDK/runtime.
public enum MobileScreen: String { case home, session, settings, memory, fontSize = "font-size", storage }
public enum AgentLocation: String, Hashable { case local = "本地工作", cloud = "云端工作", remote = "远程电脑" }
public enum ConnectionState: Hashable { case ready, offline, syncing, error }
public struct StorageItem: Identifiable, Equatable { public let id: String; public let title: String; public let detail: String; public var sizeMB: Double }
public struct MobileTask: Identifiable, Equatable, Codable { public let id: String; public let title: String; public let status: String }
public struct MobileMessage: Identifiable, Equatable { public let id = UUID(); public let role: String; public let content: String }
public struct RemoteDeviceDescriptor: Identifiable, Equatable { public let id: String; public let displayName: String; public let platform: String; public let online: Bool; public let publicKey: Data; public let lanEndpoints: [String] }

public struct MobileSurfaceState: Equatable {
    public var screen: MobileScreen = .home
    public var drawerOpen = false
    public var sheet: String?
    public var sending = false
    public var permissionGranted = false
    public var location: AgentLocation = .local
    public var connection: ConnectionState = .ready
    public var selectedModel = "Auto"
    public var theme = "system"
    public var language = "跟随系统"
    public var fontSize = 16
    public var followSystemFont = false
    public var notificationsEnabled = true
    public var memoryDraft = ""
    public var memories: [String] = []
    public var tasks: [MobileTask] = []
    public var activeTaskID: String?
    public var activeTaskTitle: String?
    public var messages: [MobileMessage] = []
    public var inputText = ""
    public var cloudURL = "http://127.0.0.1:8583"
    public var cloudUsername = ""
    public var cloudPassword = ""
    public var cloudStatus = "未连接"
    public var remoteDevices: [RemoteDeviceDescriptor] = []
    public var selectedRemoteDeviceID: String?
    public var storage: [StorageItem] = []
    public init() {}

    public static func load() -> MobileSurfaceState {
        var state = MobileSurfaceState()
        let defaults = UserDefaults.standard
        state.selectedModel = defaults.string(forKey: "notemeld.mobile.model") ?? "Auto"
        state.theme = defaults.string(forKey: "notemeld.mobile.theme") ?? "system"
        state.language = defaults.string(forKey: "notemeld.mobile.language") ?? "跟随系统"
        state.fontSize = defaults.object(forKey: "notemeld.mobile.fontSize") as? Int ?? 16
        state.followSystemFont = defaults.bool(forKey: "notemeld.mobile.followSystemFont")
        state.notificationsEnabled = defaults.object(forKey: "notemeld.mobile.notifications") as? Bool ?? true
        state.cloudURL = defaults.string(forKey: "notemeld.mobile.cloudURL") ?? "http://127.0.0.1:8583"
        state.cloudUsername = defaults.string(forKey: "notemeld.mobile.cloudUsername") ?? ""
        state.memories = defaults.stringArray(forKey: "notemeld.mobile.memories") ?? []
        if let data = defaults.data(forKey: "notemeld.mobile.tasks"), let tasks = try? JSONDecoder().decode([MobileTask].self, from: data) { state.tasks = tasks }
        state.storage = storageItems()
        return state
    }

    public func persist() {
        let defaults = UserDefaults.standard
        defaults.set(selectedModel, forKey: "notemeld.mobile.model")
        defaults.set(theme, forKey: "notemeld.mobile.theme")
        defaults.set(language, forKey: "notemeld.mobile.language")
        defaults.set(fontSize, forKey: "notemeld.mobile.fontSize")
        defaults.set(followSystemFont, forKey: "notemeld.mobile.followSystemFont")
        defaults.set(notificationsEnabled, forKey: "notemeld.mobile.notifications")
        defaults.set(cloudURL, forKey: "notemeld.mobile.cloudURL")
        defaults.set(cloudUsername, forKey: "notemeld.mobile.cloudUsername")
        defaults.set(memories, forKey: "notemeld.mobile.memories")
        if let data = try? JSONEncoder().encode(tasks) { defaults.set(data, forKey: "notemeld.mobile.tasks") }
    }

    public static func clearStorage(_ id: String?) {
        for (key, directory) in storageDirectories() where id == nil || id == key {
            try? FileManager.default.removeItem(at: directory)
            try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        }
    }

    public static func cloudAttachmentDirectory() -> URL {
        let directory = storageDirectories()["cloud"]!.appendingPathComponent("attachments", isDirectory: true)
        try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
        return directory
    }

    public static func storageItems() -> [StorageItem] {
        let directories = storageDirectories()
        return [
            StorageItem(id: "cache", title: "缓存", detail: "接口、预览与附件缓存", sizeMB: sizeMB(directories["cache"]!)),
            StorageItem(id: "cloud", title: "云端对话数据", detail: "本机保存的云端事件与缓存", sizeMB: sizeMB(directories["cloud"]!)),
            StorageItem(id: "local", title: "本机对话数据", detail: "仅保存在本机的会话记录", sizeMB: sizeMB(directories["local"]!)),
            StorageItem(id: "artifacts", title: "下载的产物文件", detail: "沙箱下载的预览与产物", sizeMB: sizeMB(directories["artifacts"]!)),
        ]
    }

    private static func storageDirectories() -> [String: URL] {
        let fileManager = FileManager.default
        let base = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("NoteMeld", isDirectory: true)
        let cache = fileManager.urls(for: .cachesDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("NoteMeld", isDirectory: true)
        let directories = [
            "cache": cache,
            "cloud": base.appendingPathComponent("cloud-cache", isDirectory: true),
            "local": base.appendingPathComponent("local", isDirectory: true),
            "artifacts": base.appendingPathComponent("artifacts", isDirectory: true),
        ]
        for directory in directories.values { try? fileManager.createDirectory(at: directory, withIntermediateDirectories: true) }
        return directories
    }

    private static func sizeMB(_ directory: URL) -> Double {
        let fileManager = FileManager.default
        guard let enumerator = fileManager.enumerator(at: directory, includingPropertiesForKeys: [.fileSizeKey], options: [.skipsHiddenFiles]) else { return 0 }
        let bytes = enumerator.compactMap { item -> Int64? in
            guard let url = item as? URL, let values = try? url.resourceValues(forKeys: [.isRegularFileKey, .fileSizeKey]), values.isRegularFile == true else { return nil }
            return Int64(values.fileSize ?? 0)
        }.reduce(0, +)
        return Double(bytes) / (1024 * 1024)
    }
}

public enum MobileSurfaceAction { case navigate(MobileScreen), toggleDrawer, openSheet(String), closeOverlay, setSending(Bool), setPermission(Bool), setLocation(AgentLocation), setConnection(ConnectionState), selectModel(String), setTheme(String), setLanguage(String), setFontSize(Int), setFollowSystemFont(Bool), setNotifications(Bool), setMemoryDraft(String), setTasks([MobileTask]), openTask(String?, String), setMemories([String]), setInputText(String), addMessage(MobileMessage), setCloudURL(String), setCloudUsername(String), setCloudPassword(String), setCloudStatus(String), setRemoteDevices([RemoteDeviceDescriptor]), selectRemoteDevice(String), addMemory, clearStorage(String?) }

public func reduceMobileSurface(_ state: MobileSurfaceState, _ action: MobileSurfaceAction) -> MobileSurfaceState {
    var next = state
    switch action {
    case let .navigate(screen): next.screen = screen; next.drawerOpen = false; next.sheet = nil
    case .toggleDrawer: next.drawerOpen.toggle(); next.sheet = nil
    case let .openSheet(name): next.sheet = name; next.drawerOpen = false
    case .closeOverlay: next.drawerOpen = false; next.sheet = nil
    case let .setSending(value): next.sending = value
    case let .setPermission(value): next.permissionGranted = value
    case let .setLocation(value): next.location = value
    case let .setConnection(value): next.connection = value
    case let .selectModel(value): next.selectedModel = value; next.sheet = nil
    case let .setTheme(value): next.theme = value; next.sheet = nil
    case let .setLanguage(value): next.language = value; next.sheet = nil
    case let .setFontSize(value): next.fontSize = value
    case let .setFollowSystemFont(value): next.followSystemFont = value
    case let .setNotifications(value): next.notificationsEnabled = value
    case let .setMemoryDraft(value): next.memoryDraft = value
    case let .setTasks(value): next.tasks = value
    case let .openTask(id, title): next.activeTaskID = id; next.activeTaskTitle = title; next.screen = .session; next.drawerOpen = false
    case let .setMemories(value): next.memories = value
    case let .setInputText(value): next.inputText = value
    case let .addMessage(value): next.messages.append(value)
    case let .setCloudURL(value): next.cloudURL = value
    case let .setCloudUsername(value): next.cloudUsername = value
    case let .setCloudPassword(value): next.cloudPassword = value
    case let .setCloudStatus(value): next.cloudStatus = value
    case let .setRemoteDevices(value): next.remoteDevices = value; next.selectedRemoteDeviceID = value.first(where: { $0.online })?.id ?? value.first?.id
    case let .selectRemoteDevice(value): next.selectedRemoteDeviceID = value
    case .addMemory: if !next.memoryDraft.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { next.memories.append(next.memoryDraft.trimmingCharacters(in: .whitespacesAndNewlines)); next.memoryDraft = "" }
    case let .clearStorage(id): if let id { next.storage.removeAll { $0.id == id } } else { next.storage.removeAll() }
    }
    return next
}

File: android/src/main/kotlin/com/notemeld/mobile/MobileSurfaceState.kt
package com.notemeld.mobile

/** Shared page contract for the Android host. Agent execution stays in the SDK/runtime. */
enum class MobileScreen { HOME, SESSION, SETTINGS, MEMORY, FONT_SIZE, STORAGE }
enum class AgentLocation { LOCAL, CLOUD, REMOTE }
enum class ConnectionState { READY, OFFLINE, SYNCING, ERROR }

data class StorageItem(val id: String, val title: String, val description: String, val sizeMb: Double)
data class MobileTask(val id: String, val title: String, val status: String)
data class MobileMessage(val role: String, val content: String)
data class MobileAttachment(val name: String, val contentBase64: String, val size: Int, val type: String)
data class RemoteDevice(val id: String, val displayName: String, val platform: String, val online: Boolean, val publicKey: String? = null, val lanEndpoints: List<String> = emptyList())

data class MobileSurfaceState(
    val screen: MobileScreen = MobileScreen.HOME,
    val drawerOpen: Boolean = false,
    val sheet: String? = null,
    val sending: Boolean = false,
    val permissionGranted: Boolean = false,
    val location: AgentLocation = AgentLocation.LOCAL,
    val connection: ConnectionState = ConnectionState.READY,
    val selectedModel: String = "Auto",
    val theme: String = "system",
    val language: String = "跟随系统",
    val fontSize: Int = 16,
    val followSystemFont: Boolean = false,
    val notificationsEnabled: Boolean = true,
    val memoryDraft: String = "",
    val inputText: String = "",
    val cloudUrl: String = "http://10.0.2.2:8583",
    val cloudUsername: String = "",
    val cloudPassword: String = "",
    val cloudStatus: String = "未连接",
    val remoteDevices: List<RemoteDevice> = emptyList(),
    val selectedRemoteDeviceId: String? = null,
    val showActionMenu: Boolean = false,
    val toast: String? = null,
    val selectedWorkTab: AgentLocation = AgentLocation.LOCAL,
    val attachments: List<String> = emptyList(),
    val tasks: List<MobileTask> = emptyList(),
    val activeTaskId: String? = null,
    val activeTaskTitle: String? = null,
    val messages: List<MobileMessage> = emptyList(),
    val memories: List<String> = emptyList(),
    val storage: List<StorageItem> = emptyList(),
)

sealed interface MobileSurfaceAction {
    data class Navigate(val screen: MobileScreen) : MobileSurfaceAction
    data object ToggleDrawer : MobileSurfaceAction
    data class OpenSheet(val name: String) : MobileSurfaceAction
    data object CloseOverlay : MobileSurfaceAction
    data class SetSending(val value: Boolean) : MobileSurfaceAction
    data class SetPermission(val value: Boolean) : MobileSurfaceAction
    data class SetLocation(val value: AgentLocation) : MobileSurfaceAction
    data class SetConnection(val value: ConnectionState) : MobileSurfaceAction
    data class SelectModel(val value: String) : MobileSurfaceAction
    data class SetTheme(val value: String) : MobileSurfaceAction
    data class SetLanguage(val value: String) : MobileSurfaceAction
    data class SetFontSize(val value: Int) : MobileSurfaceAction
    data class SetFollowSystemFont(val value: Boolean) : MobileSurfaceAction
    data class SetNotifications(val value: Boolean) : MobileSurfaceAction
    data class SetMemoryDraft(val value: String) : MobileSurfaceAction
    data class SetInputText(val value: String) : MobileSurfaceAction
    data object ToggleActionMenu : MobileSurfaceAction
    data class AddAttachment(val name: String) : MobileSurfaceAction
    data object ShowProducts : MobileSurfaceAction
    data object DeleteConversation : MobileSurfaceAction
    data class ShowToast(val value: String) : MobileSurfaceAction
    data class SetWorkTab(val value: AgentLocation) : MobileSurfaceAction
    data object AddMemory : MobileSurfaceAction
    data class SetMemories(val value: List<String>) : MobileSurfaceAction
    data class SetTasks(val value: List<MobileTask>) : MobileSurfaceAction
    data class OpenTask(val id: String?, val title: String) : MobileSurfaceAction
    data class AddMessage(val value: MobileMessage) : MobileSurfaceAction
    data class SetCloudUrl(val value: String) : MobileSurfaceAction
    data class SetCloudUsername(val value: String) : MobileSurfaceAction
    data class SetCloudPassword(val value: String) : MobileSurfaceAction
    data class SetCloudStatus(val value: String) : MobileSurfaceAction
    data class SetRemoteDevices(val value: List<RemoteDevice>) : MobileSurfaceAction
    data class SelectRemoteDevice(val value: String) : MobileSurfaceAction
    data class ClearStorage(val id: String?) : MobileSurfaceAction
}

fun reduceMobileSurface(state: MobileSurfaceState, action: MobileSurfaceAction): MobileSurfaceState = when (action) {
    is MobileSurfaceAction.Navigate -> state.copy(screen = action.screen, drawerOpen = false, sheet = null)
    MobileSurfaceAction.ToggleDrawer -> state.copy(drawerOpen = !state.drawerOpen, sheet = null)
    is MobileSurfaceAction.OpenSheet -> state.copy(sheet = action.name, drawerOpen = false)
    MobileSurfaceAction.CloseOverlay -> state.copy(drawerOpen = false, sheet = null)
    is MobileSurfaceAction.SetSending -> state.copy(sending = action.value)
    is MobileSurfaceAction.SetPermission -> state.copy(permissionGranted = action.value)
    is MobileSurfaceAction.SetLocation -> state.copy(location = action.value)
    is MobileSurfaceAction.SetConnection -> state.copy(connection = action.value)
    is MobileSurfaceAction.SelectModel -> state.copy(selectedModel = action.value, sheet = null)
    is MobileSurfaceAction.SetTheme -> state.copy(theme = action.value, sheet = null)
    is MobileSurfaceAction.SetLanguage -> state.copy(language = action.value, sheet = null)
    is MobileSurfaceAction.SetFontSize -> state.copy(fontSize = action.value)
    is MobileSurfaceAction.SetFollowSystemFont -> state.copy(followSystemFont = action.value)
    is MobileSurfaceAction.SetNotifications -> state.copy(notificationsEnabled = action.value)
    is MobileSurfaceAction.SetMemoryDraft -> state.copy(memoryDraft = action.value)
    is MobileSurfaceAction.SetInputText -> state.copy(inputText = action.value)
    MobileSurfaceAction.ToggleActionMenu -> state.copy(showActionMenu = !state.showActionMenu)
    is MobileSurfaceAction.AddAttachment -> state.copy(attachments = state.attachments + action.name)
    MobileSurfaceAction.ShowProducts -> state.copy(showActionMenu = false, sheet = "products")
    MobileSurfaceAction.DeleteConversation -> state.copy(showActionMenu = false, screen = MobileScreen.HOME)
    is MobileSurfaceAction.ShowToast -> state.copy(toast = action.value)
    is MobileSurfaceAction.SetWorkTab -> state.copy(selectedWorkTab = action.value)
    MobileSurfaceAction.AddMemory -> if (state.memoryDraft.isBlank()) state else state.copy(
        memories = state.memories + state.memoryDraft.trim(), memoryDraft = ""
    )
    is MobileSurfaceAction.SetMemories -> state.copy(memories = action.value)
    is MobileSurfaceAction.SetTasks -> state.copy(tasks = action.value)
    is MobileSurfaceAction.OpenTask -> state.copy(activeTaskId = action.id, activeTaskTitle = action.title, screen = MobileScreen.SESSION, drawerOpen = false)
    is MobileSurfaceAction.AddMessage -> state.copy(messages = state.messages + action.value)
    is MobileSurfaceAction.SetCloudUrl -> state.copy(cloudUrl = action.value)
    is MobileSurfaceAction.SetCloudUsername -> state.copy(cloudUsername = action.value)
    is MobileSurfaceAction.SetCloudPassword -> state.copy(cloudPassword = action.value)
    is MobileSurfaceAction.SetCloudStatus -> state.copy(cloudStatus = action.value)
    is MobileSurfaceAction.SetRemoteDevices -> state.copy(remoteDevices = action.value, selectedRemoteDeviceId = action.value.firstOrNull { it.online }?.id ?: action.value.firstOrNull()?.id)
    is MobileSurfaceAction.SelectRemoteDevice -> state.copy(selectedRemoteDeviceId = action.value)
    is MobileSurfaceAction.ClearStorage -> state
}

fun reduceAndPersistMobileSurface(context: android.content.Context, state: MobileSurfaceState, action: MobileSurfaceAction): MobileSurfaceState {
    val next = reduceMobileSurface(state, action)
    val persisted = if (action is MobileSurfaceAction.ClearStorage) {
        next.copy(storage = MobileLocalStore.clear(context, action.id))
    } else next
    MobileLocalStore.save(context, persisted)
    return persisted
}
